azure/converters/vmss.go

```diff
@@ -46,25 +45,65 @@ func SDKToVMSS(sdkvmss compute.VirtualMachineScaleSet, sdkinstances []compute.Vi
 	}
 
 	if len(sdkinstances) > 0 {
-		vmss.Instances = make([]infrav1exp.VMSSVM, len(sdkinstances))
+		vmss.Instances = make([]azure.VMSSVM, len(sdkinstances))
 		for i, vm := range sdkinstances {
-			instance := infrav1exp.VMSSVM{
-				ID:         to.String(vm.ID),
-				InstanceID: to.String(vm.InstanceID),
-				Name:       to.String(vm.OsProfile.ComputerName),
-				State:      infrav1.ProvisioningState(to.String(vm.ProvisioningState)),
-			}
-
-			if vm.LatestModelApplied != nil {
-				instance.LatestModelApplied = *vm.LatestModelApplied
-			}
-
-			if vm.Zones != nil && len(*vm.Zones) > 0 {
-				instance.AvailabilityZone = to.StringSlice(vm.Zones)[0]
-			}
-			vmss.Instances[i] = instance
+			vmss.Instances[i] = *SDKToVMSSVM(vm)
 		}
 	}
 
+	if sdkvmss.VirtualMachineProfile != nil &&
+		sdkvmss.VirtualMachineProfile.StorageProfile != nil &&
+		sdkvmss.VirtualMachineProfile.StorageProfile.ImageReference != nil {
+
+		imageRef := sdkvmss.VirtualMachineProfile.StorageProfile.ImageReference
+		vmss.Image = SDKImageToImage(imageRef)
+	}
+
 	return vmss
 }
+
+// SDKToVMSSVM converts an Azure SDK VirtualMachineScaleSetVM into an infrav1exp.VMSSVM
+func SDKToVMSSVM(sdkInstance compute.VirtualMachineScaleSetVM) *azure.VMSSVM {
+	instance := azure.VMSSVM{
+		ID:         to.String(sdkInstance.ID),
+		InstanceID: to.String(sdkInstance.InstanceID),
+	}
+
+	if sdkInstance.VirtualMachineScaleSetVMProperties == nil {
+		return &instance
+	}
+
+	instance.State = infrav1.Creating
+	if sdkInstance.ProvisioningState != nil {
+		instance.State = infrav1.ProvisioningState(to.String(sdkInstance.ProvisioningState))
+	}
+
+	if sdkInstance.OsProfile != nil && sdkInstance.OsProfile.ComputerName != nil {
+		instance.Name = *sdkInstance.OsProfile.ComputerName
+	}
+
+	if sdkInstance.StorageProfile != nil && sdkInstance.StorageProfile.ImageReference != nil {
+		imageRef := sdkInstance.StorageProfile.ImageReference
+		instance.Image = SDKImageToImage(imageRef)
+	}
+
+	if sdkInstance.Zones != nil && len(*sdkInstance.Zones) > 0 {
+		instance.AvailabilityZone = to.StringSlice(sdkInstance.Zones)[0]
```

Comment:
why is this hardcoded to zero? Do we expect any cases where there is more than 1 zone there? If not, can you please add a code comment to indicate that?

Reply:
Instance can only be in 1 zone. I believe the same thing happens for VMs. I'll add a comment.